src/utils/ftexpression.rs

```diff
@@ -0,0 +1,194 @@
+        if operator != "=" {
+            eprintln!("Range thresholds can only be used with the '=' operator.");
+            std::process::exit(1);
+        }
+    }
+
+    let threshold_value = match range {
+        Some((min, max)) => Threshold::Range(min, max),
+        None => Threshold::Single(threshold.unwrap()),
+    };
+
+    ParsedExpr {
+        fn_name: func_name,
+        rng_name: gnm_feat,
+        op: operator,
+        threshold: threshold_value,
+    }
+}
+
+pub fn apply_filter_to_range(
+    parsed: &ParsedExpr,
+    mut range: bamranges::Ranges,
+) -> Result<bamranges::Ranges, anyhow::Error> {
+    let starting_len = range.starts.len();
+
+    let to_keep: Vec<bool> = if parsed.fn_name == "len" {
+        range
+            .lengths
+            .iter()
+            .map(|l| len(l.unwrap(), &parsed.op, &parsed.threshold))
+            .collect()
+    } else if parsed.fn_name == "qual" {
+        range
+            .qual
+            .iter()
+            .map(|q| qual(*q, &parsed.op, &parsed.threshold))
+            .collect()
+    } else {
+        anyhow::bail!("Invalid function name: {}", &parsed.fn_name);
+    };
+
+    // drop i64 values from the range
+    for vec in [
+        &mut range.starts,
+        &mut range.ends,
+        &mut range.lengths,
+        &mut range.reference_starts,
+        &mut range.reference_ends,
+        &mut range.reference_lengths,
+    ] {
+        *vec = vec
+            .iter()
+            .zip(to_keep.iter())
+            .filter_map(|(v, d)| if *d { Some(*v) } else { None })
+            .collect();
+    }
+
+    // drop u8 values from the range
+    range.qual = range
+        .qual
+        .iter()
+        .zip(to_keep.iter())
+        .filter_map(|(v, d)| if *d { Some(*v) } else { None })
+        .collect();
+
+    // check we dropped the right number of values
+    let n_dropped = to_keep.iter().filter(|&&x| !x).count();
+    assert_eq!(starting_len, range.starts.len() + n_dropped);
+
+    Ok(range)
+}
+
+/// tests
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::utils::bamranges;
+
+    fn make_fake_range() -> bamranges::Ranges {
+        bamranges::Ranges {
+            starts: vec![Some(0), Some(10), Some(17)],
+            ends: vec![Some(5), Some(15), Some(20)],
+            lengths: vec![Some(5), Some(5), Some(3)],
+            qual: vec![0, 255, 181],
+            reference_starts: vec![Some(0), Some(10), Some(17)],
+            reference_ends: vec![Some(5), Some(15), Some(20)],
+            reference_lengths: vec![Some(5), Some(5), Some(3)],
+            seq_len: 100,
+            reverse: false,
+        }
+    }
+
+    #[test]
+    fn test_this_one() {
+        let filter = "len(msp)=50:100";
+        let mut range = make_fake_range();
+        let parser = parse_filter(&filter);
+        eprintln!("{:?}", range.starts.len());
+        apply_filter_to_range(&parser, &mut range).unwrap();
+        eprintln!("{:?}", range.starts.len());
```

Comment:
Please update this test so that it works as a test for your new code.

Reply:
[50f1012](https://github.com/fiberseq/fibertools-rs/commit/50f101278debedaef50ef82dc39db1d72c565734)